### PR TITLE
[Swift 6] `Sendable` conformance to allow defining actor-isolated `Factory`s

### LINKED ISSUE #215: Swift 6 concurrency warnings
Given containers like such on Factory `2.3.2`:

```swift
public final class CoreContainer: SharedContainer {
     public static let shared = CoreContainer()
     public let manager = ContainerManager()
}
```

With Swift 6 complete concurrency we get the following error `Class property 'shared' is not concurrency-safe because non-'Sendable' type 'CoreContainer' may have shared mutable state`.

Is Factory up-to-date with concurrency yet?

### PR DESCRIPTION
**Pull request by NachoSoto**
Fixes #215

This allows doing this:
```swift
private struct SomeSendableType: Sendable {}

extension Container {
  @MainActor
  var instance: Factory<SomeSendableType> {
    self { .init() }
  }
}
```

### CODE REVIEW

**Comment by NachoSoto** [comment]
Also note that without this, users might run into https://github.com/swiftlang/swift/issues/75453 and crash at runtime.

File: Tests/FactoryTests/FactoryIsolationTests.swift [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+final class FactoryIsolationTests: XCTestCase {
+
+  override func setUp() {
+    super.setUp()
+    Container.shared.reset()
+  }
+
+  func testInjectMainActorDependency() async {
+    let _: SomeSendableType = await Container.shared.instance()
```
Comment (NachoSoto) [inline Tests/FactoryTests/FactoryIsolationTests.swift:21]: `await` to be able to call the `@MainActor` factory.

File: Tests/FactoryTests/FactoryIsolationTests.swift [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    super.setUp()
+    Container.shared.reset()
+  }
+
+  func testInjectSendableDependency() {
+    let _: SomeSendableType = Container.shared.sendable()
+  }
+
+  func testInjectMainActorDependency() async {
+    let _: SomeMainActorType = await Container.shared.mainActor().wrapped()
```
Comment (NachoSoto) [inline Tests/FactoryTests/FactoryIsolationTests.swift:42]: I took this example from https://github.com/hmlongco/Factory/commit/3ca72ff47173435b85e1df40b1dee8c34e6ab57e but this isn't viable... `Factory` would be caching `AsyncInitWrapper` but not the actual instance created by it.

Comment (NachoSoto) [reply Tests/FactoryTests/FactoryIsolationTests.swift:42]: Unfortunately I can't think of a solution to this given that Swift (6) does not have a way to parameterize over an Actor.

Comment (corujautx) [reply Tests/FactoryTests/FactoryIsolationTests.swift:42]: @NachoSoto have you had a look into [this macro](https://developer.apple.com/documentation/swift/isolation()) ?
Sorry, still trying to wrap my head around Swift Concurrency, so might be not related.

**Comment by ulian-onua** [comment]
@NachoSoto with this change, I get an error if try to register @MainActor class using Factory like this:

`Call to main actor-isolated initializer 'init(dataSource:)' in a synchronous nonisolated context`

Any ideas?

**Comment by hmlongco** [comment]
@ulian-onua Checkout the concurrency branch.

https://github.com/hmlongco/Factory/tree/concurrency

**Comment by ulian-onua** [comment]
@hmlongco thank you. Is it planned to merge this `concurrency` branch into `main` and release then?

**Comment by hmlongco** [comment]
@ulian-onua It is, but I put it out as a test branch first in order to make sure that approach wasn't going to cause even more problems.

**Comment by ulian-onua** [comment]
@hmlongco ok. Thank you. Please solve it as soon as possible. A lot of people use your library and it is strange to get this major change that breaks the whole app. Thank you.

**Comment by gdetari** [comment]
I am using the concurrency branch, and I still have this exact error.

**Comment by hmlongco** [comment]
The concurrency branch is no longer current, and @MainActor support is in main. Note if you're defining such you need to specify the actor context in the factory closure.
```swift
extension Container {
    @MainActor
    fileprivate var mainActor: Factory<SomeMainActorType> {
        self { @MainActor in SomeMainActorType() }
    }
}
```

**Pull request merged**